### PR TITLE
Add view support for iceberg jdbc catalog

### PR DESCRIPTION
**Pull request by tbaeg**
## Description
Add view support for iceberg JDBC catalog.

## Additional context and related issues
Iceberg now supports views. The iceberg REST catalog already supports views. This just brings things to parity.

## Release notes

```markdown
# Iceberg
* Add support for views in JDBC catalog. ({issue}`22576`)
* Upgrade `jdbc.schema-version` to `V1` in JDBC catalog. ({issue}`22576`)
```

### CODE REVIEW

**Comment by tbaeg** [comment]
I don't believe the failure is related.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java [hunk 1]
```diff
@@ -398,19 +446,48 @@ public void setViewPrincipal(ConnectorSession session, SchemaTableName schemaVie
...
 
     @Override
     public Map<SchemaTableName, ConnectorViewDefinition> getViews(ConnectorSession session, Optional<String> namespace)
     {
-        return ImmutableMap.of();
+        ImmutableMap.Builder<SchemaTableName, ConnectorViewDefinition> views = ImmutableMap.builder();
+        for (Namespace ns : jdbcCatalog.listNamespaces()) {
+            for (TableIdentifier restView : jdbcCatalog.listViews(ns)) {
+                SchemaTableName schemaTableName = SchemaTableName.schemaTableName(restView.namespace().toString(), restView.name());
+                getView(session, schemaTableName).ifPresent(view -> views.put(schemaTableName, view));
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java:459]: This logic may throw an exception if views with unsupported dialect exist in the JDBC catalog. 

Please take a look at REST catalog implementation & `TestIcebergTrinoRestCatalogConnectorSmokeTest#testUnsupportedViewDialect`:
https://github.com/trinodb/trino/blob/facddc80df514c0aa05341d74d0cb62004b63d8d/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoRestCatalog.java#L444-L453

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalogFactory.java [hunk 2]
```diff
@@ -63,14 +64,16 @@ public TrinoJdbcCatalogFactory(
         this.typeManager = requireNonNull(typeManager, "typeManager is null");
         this.tableOperationsProvider = requireNonNull(tableOperationsProvider, "tableOperationsProvider is null");
         this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
-        this.isUniqueTableLocation = requireNonNull(icebergConfig, "icebergConfig is null").isUniqueTableLocation();
         this.jdbcClient = requireNonNull(jdbcClient, "jdbcClient is null");
         this.jdbcCatalogName = jdbcConfig.getCatalogName();
         this.defaultWarehouseDir = jdbcConfig.getDefaultWarehouseDir();
+        this.schemaVersion = jdbcConfig.getSchemaVersion();
+        this.isUniqueTableLocation = requireNonNull(icebergConfig, "icebergConfig is null").isUniqueTableLocation();
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalogFactory.java:71]: Remove redundant `requireNonNull`.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalogFactory.java [hunk 3]
```diff
@@ -45,6 +45,7 @@ public class TrinoJdbcCatalogFactory
     private final IcebergJdbcClient jdbcClient;
     private final String jdbcCatalogName;
     private final String defaultWarehouseDir;
+    private final String schemaVersion;
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalogFactory.java:48]: Convert to a local variable.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalogFactory.java [hunk 4]
```diff
@@ -88,8 +91,8 @@ public void shutdown()
     public TrinoCatalog create(ConnectorIdentity identity)
     {
         JdbcCatalog jdbcCatalog = new JdbcCatalog(
-                config -> new ForwardingFileIo(fileSystemFactory.create(identity)),
-                config -> clientPool,
+                _ -> new ForwardingFileIo(fileSystemFactory.create(identity)),
+                _ -> clientPool,
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalogFactory.java:95]: Revert unrelated changes.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogConfig.java [hunk 5]
```diff
@@ -23,12 +23,16 @@
 
 public class IcebergJdbcCatalogConfig
 {
+    public static final String ICEBERG_JDBC_CATALOG_SCHEMA_V0 = "V0";
+    public static final String ICEBERG_JDBC_CATALOG_SCHEMA_V1 = "V1";
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogConfig.java:27]: You could add an enum instead.

**Comment by tbaeg** [comment]
@ebyhr Thanks for the review! Updated the PR relative to your feedback. One thing I did notice is adding the `testUnsupportedViewDialect` test showed I needed to return discrete tables and views. It made me wonder if views that are not using the supported dialect should even show up in a `SHOW TABLES` call. This appears to be the behavior for the REST catalog as well. Thoughts?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogConfig.java (hunk 5 above)
Comment (tbaeg) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogConfig.java:30]: Updated.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogConfig.java [hunk 6]
```diff
@@ -23,12 +23,18 @@
...
+        V0, V1
+    }
+
     private String driverClass;
     private String connectionUrl;
     private String connectionUser;
     private String connectionPassword;
     private String catalogName;
     private String defaultWarehouseDir;
+    private String schemaVersion = SchemaVersion.V1.toString();
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogConfig.java:37]: Why not use enum directly?

**Comment by tbaeg** [comment]
Checking in to see if there were any other changes required to get this merged, thanks!

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java [hunk 7]
```diff
@@ -25,14 +26,25 @@ public class IcebergJdbcClient
...
         this.catalogName = requireNonNull(catalogName, "catalogName is null");
+        this.schemaVersion = requireNonNull(schemaVersion, "schemaVersion is null");
     }
 
     public void createTable(String schemaName, String tableName, String metadataLocation)
+    {
+        switch (schemaVersion) {
+            case SchemaVersion.V0 -> createTableV0(schemaName, tableName, metadataLocation);
+            case SchemaVersion.V1 -> createTableV1(schemaName, tableName, metadataLocation);
+            default -> throw new IllegalArgumentException("Unsupported schema version: " + schemaVersion);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java:43]: Please remove this default clause. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-default-clause-in-exhaustive-enum-based-switch-statements

Comment (tbaeg) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java]: Done!

**Comment by ebyhr** [comment]
Could you rebase on master to resolve conflicts?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcCatalogConnectorSmokeTest.java [hunk 8]
```diff
@@ -117,11 +122,32 @@ public final void destroy()
     }
 
     @Test
-    @Override
-    public void testView()
+    void testUnsupportedViewDialect()
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcCatalogConnectorSmokeTest.java:125]: We found a query failure when executing `DROP SCHEMA ... CASCADE` against schemas with views #22758 in REST catalog. Could you add a similar test and fix it if necessary?

Comment (tbaeg) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcCatalogConnectorSmokeTest.java:157]: Done!

**Comment by ebyhr** [comment]
Could you confirm CI failures?

**Comment by tbaeg** [comment]
> Could you confirm CI failures?

Looks one failure related to `MissingCasesInEnumSwitch`. All enum cases are handled but seems like the plugin just doesn't like the qualified reference. I can remove the qualifer.

Everything else does not seem related.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java [hunk 9]
```diff
@@ -47,6 +60,22 @@ public void createTable(String schemaName, String tableName, String metadataLoca
...
+    {
+        try (Handle handle = Jdbi.open(connectionFactory)) {
+            handle.createUpdate("" +
+                            "INSERT INTO iceberg_tables " +
+                            "(catalog_name, table_namespace, table_name, metadata_location, previous_metadata_location, iceberg_type) " +
+                            "VALUES (:catalog, :schema, :table, :metadata_location, null, :iceberg_type)")
+                    .bind("catalog", catalogName)
+                    .bind("schema", schemaName)
+                    .bind("table", tableName)
+                    .bind("iceberg_type", "TABLE")
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java:73]: Can we inline this constant in INSERT statement?

Comment (tbaeg) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java]: Done!

**Comment by tbaeg** [comment]
CI failure is not related. @ebyhr

**Review by ebyhr — APPROVED** [review]
